cmd/routesrv/main.go

```diff
@@ -0,0 +1,279 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+	"time"
+
+	ot "github.com/opentracing/opentracing-go"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	log "github.com/sirupsen/logrus"
+	"github.com/zalando/skipper/cmd/routesrv/options"
+	"github.com/zalando/skipper/config"
+	"github.com/zalando/skipper/dataclients/kubernetes"
+	"github.com/zalando/skipper/eskip"
+	"github.com/zalando/skipper/routing"
+	"github.com/zalando/skipper/tracing"
+)
+
+// eskipBytes keeps eskip-formatted routes as a byte slice and
+// provides synchronized r/w access to them. Additionally it can
+// serve as an HTTP handler exposing its content.
+type eskipBytes struct {
+	data []byte
+	mu   sync.RWMutex
+
+	tracer ot.Tracer
+}
+
+func (e *eskipBytes) bytes() []byte {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+
+	return e.data
+}
+
+// formatAndSet takes a slice of routes and stores them eskip-formatted
+// in a synchronized way. References to both new and old data are returned
+// for inspection (reading). Returned slices content must not be modified.
+func (e *eskipBytes) formatAndSet(routes []*eskip.Route) ([]byte, []byte) {
+	buf := &bytes.Buffer{}
+	eskip.Fprint(buf, eskip.PrettyPrintInfo{}, routes...)
+
+	e.mu.Lock()
+	oldData := e.data
+	e.data = buf.Bytes()
+	e.mu.Unlock()
+
+	return e.data, oldData
+}
+
+func (e *eskipBytes) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	span := tracing.CreateSpan("serve_routes", r.Context(), e.tracer)
+	defer span.Finish()
+
+	if data := e.bytes(); data != nil {
```

Comment:
i just noticed the earlier suggestion from @AlexanderYastrebov . But in Go, comparing slices to nil is not a good practice. The reason is that a lot of existing functions handle nil and []byte{} the same way, plus, handling nil and empty differently would make it necessary when writing new functions in our code to remember that we need to return empty instead of nil in this program.

Besides that, in this case we also carry here an important meaning: uninitialized. So, it's better to be explicit.

Reply:
> i just noticed the earlier suggestion

Well, maybe not the best one :) Lets make it explicit, I agree